File: tools/dom-xss-analyzer/examples/vulnerable.java
// Java/Servlet + Spring example for dxa. Every pattern below is intentionally
// unsafe and is for detector testing only.
package com.example;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.web.bind.annotation.*;

public class VulnerableController {

    // HIGH: servlet request source flows straight into response.getWriter
    public void doGet(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String q = request.getParameter("q");
        response.getWriter().write("<div>You searched: " + q + "</div>");
    }

    // HIGH: taint propagates through a local variable, then hits print
    public void handle(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String name = request.getHeader("X-User-Name");
        String greet = "Hello " + name;
        response.getWriter().println(greet);
    }

    // HIGH: Spring @RequestParam source concat'd into response body
    @GetMapping(value = "/search", produces = "text/html")
    @ResponseBody
    public String search(@RequestParam String q, HttpServletResponse response) throws Exception {
        response.getWriter().print("<h1>Results for " + q + "</h1>");
        return "";
    }

    // HIGH: @PathVariable reaches a servlet write
    @GetMapping("/user/{id}/greet")
    public void greet(@PathVariable String id, HttpServletResponse response) throws Exception {
        response.getWriter().write("Welcome user " + id);
    }

    // MEDIUM: jsoup Element.html(x) sink with dynamic value
    public void loadFragment(String rawHtml) {
        org.jsoup.nodes.Element div = new org.jsoup.nodes.Element("div");
        div.html(rawHtml);
    }
}

File: tools/dom-xss-analyzer/examples/vulnerable.py
# Python example for dxa (Flask + FastAPI + Django patterns). Every pattern
# below is intentionally unsafe and is for detector testing only.
from flask import Flask, request, render_template_string, Markup, Response, make_response
from fastapi import FastAPI, Query, Header
from fastapi.responses import HTMLResponse
from django.utils.safestring import mark_safe

app = Flask(__name__)
fapi = FastAPI()


# HIGH: request source flows straight into a render_template_string() call
@app.get("/search")
def search():
    q = request.args.get("q", "")
    return render_template_string("<h1>Results for " + q + "</h1>")


# HIGH: taint propagates through a local -> Markup() sink
@app.get("/user")
def user_page():
    name = request.args["name"]
    greeting = "Hello " + name
    return Markup(greeting)


# HIGH: Django mark_safe on an unescaped value
def render_greeting(request):
    who = request.GET.get("who", "")
    return mark_safe("<div>Welcome " + who + "</div>")


# HIGH: FastAPI HTMLResponse with a query param dropped in raw
@fapi.get("/hi", response_class=HTMLResponse)
def hi(q: str = Query("")):
    return "<p>hi " + q + "</p>"


# MEDIUM: Flask make_response with text/html and dynamic body
@app.get("/note")
def note():
    body = request.form.get("body", "")
    resp = make_response("<div>" + body + "</div>")
    resp.mimetype = "text/html"
    return resp


# HIGH: Jinja |safe filter used on a variable
_TEMPLATE = "<article>{{ post|safe }}</article>"


# MEDIUM: os.system with a user value - not XSS but Python-only injection worth flagging
def run(request):
    filename = request.args.get("f", "")
    import os
    os.system("ls " + filename)
